File: NLP_Multiprocessing_For_Preprocessing_Experiment.R

#load necessary libraries
library(csv)
library(future.apply)
library(ggplot2)
library(parallel)

print("Number of logical cores:")
print(detectCores(logical = TRUE))
print("Number of physical cores:")
print(detectCores(logical = FALSE))


#read in data
all_data = read.csv("./data/news_data_Experimentation.csv", sep = ";")
#read in stopwords
all_stopwords = read.delim("./data/stopwords.txt", sep = "\n", skip = 3)
all_stopwords = all_stopwords[,1]
all_stopwords = as.vector(all_stopwords)
stopwords = all_stopwords

remove_stop_words = function(text) {
  # A simple function that first splits text via
  # whitespaces and remove then stopwords that appear in list

  # split text by whitespace
  x = unlist(strsplit(text, " "))
  # check if word is in  stopwords
  x = x[!x %in% stopwords]
  new_text = paste(x, collapse = " ")

  # if text is longer than 5 characters return modified text
  # else return original text
  if (nchar(new_text) > 5) {
    return(new_text)
  }else{
    return(text)
    }
}

measure_time = function(code_chunk) {

  start_time = Sys.time()
  eval(parse(text=code_chunk))
  end_time = Sys.time()

  execution_time = end_time - start_time
  execution_time = as.numeric(execution_time, units="secs")
  print(execution_time)
}

number_of_stopwords_to_try =  c(100, 2070)
number_of_articles_to_try =  c(200, 10000, 500000)

for(n_texts in number_of_articles_to_try) {

  # select desired data
  data = all_data[1:n_texts,]

  for (n_stopwords in number_of_stopwords_to_try) {

    stopwords = all_stopwords[1:n_stopwords]

    model_name_list = c()
    execution_time_list = c()

    model_name_list = append(model_name_list, "Sequential")
    execution_time_list = append(execution_time_list,
                                 measure_time("lapply(data$text, remove_stop_words)"))

    for (n_cores in c(2,4,8)) {
      plan(multiprocess, workers = n_cores)
      model_name_list = append(model_name_list, paste("Future ", n_cores, "C.", sep = ""))
      execution_time_list = append(execution_time_list,
                                   measure_time("future_lapply(data$text, remove_stop_words)"))
    }


    #parallel with nc.cores > 1 only works on non Windows machines
    for (n_cores in c(2,4,8)) {
      model_name_list = append(model_name_list, paste("Parallel ", n_cores, "C.", sep = ""))
      execution_time_list = append(execution_time_list,
                                   measure_time("mclapply(X = data$text, FUN = remove_stop_words, mc.cores=n_cores)"))
    }

    visuals_df = data.frame(Model = model_name_list,
                            Time = execution_time_list)

    #create a plot of the results
    results_plot = ggplot(visuals_df, aes(x = Model, y= Time))+
      geom_bar(stat="identity",  fill = "light blue") +
      scale_x_discrete(labels = model_name_list) +
      labs(title="Multiprocessing Strategies in R",
           subtitle = paste("No. of texts:", n_texts,
                           "No. of Stopwords: ", n_stopwords)) +
      xlab("Multiprocessing Approach") +
      geom_hline(yintercept = min(execution_time_list), color ="red") +
      ylab("Execution Time in Seconds") +
      theme_light() +
      theme(axis.text.x = element_text(angle = 30, vjust = 0.5),
                           text = element_text(size = 20), axis.text =element_text(size = 16))
    
    ggsave(paste("./data/images/Multiprocessing_Experiment_R_",
             n_texts, "_articles_",
             n_stopwords, "_stopwords.png", sep = ""), 
           plot = results_plot)
    
    # store the results additionally as table 
    
    # adding information on experiments
    execution_time_list = c(n_texts, n_stopwords, execution_time_list)
    model_name_list = c("number_of_texts", "number_of_stopwords", model_name_list)
    
    # check if dataframe exists
    if (exists("results_df")) {
      new_df = t(data.frame(execution_time_list))
      results_df = rbind(results_df, new_df)
    
    }else{
      results_df = t(data.frame(execution_time_list))
      colnames(results_df) = model_name_list
    }
    
  }

}

write.table(results_df, "./data/Preprocessing_Experiment_Results_R.csv", 
          row.names=FALSE, quote=FALSE, sep =";")

File: NLP_Multiprocessing_For_Preprocessing_Experiment.py
import pandas as pd
import time
import timeit
import numpy as np

from joblib import Parallel, delayed
import multiprocessing as mp

import ray
import ray.util.multiprocessing

# for the visuals
import matplotlib.pyplot as plt

print(mp.cpu_count())

# read in the data set
all_data = pd.read_csv('./Data/news_data_Experimentation.csv', sep=";")

# read in stop word list
stopwords_list = []
with open('./Data/stopwords.txt') as f:
    for line in f.readlines():
        if str(line)[0] != "#":
            stopwords_list.append(str(line).replace("\n", ""))
original_stopwords_list = stopwords_list


def remove_stop_words(text):
    """ A simple function that first splits text via
    whitespaces and remove then stopwords that appear in list"""

    # split text by whitespace
    x = np.array(text.split(" "))
    # check if word is in  stopwords
    x = x[np.isin(x, stopwords_list) == False]
    # join again
    new_text = ' '.join(x.tolist())

    # if text is longer than 5 characters return modified text
    # else return original text
    if len(new_text) > 5:
        return new_text
    else:
        return text



def measure_time(name, code_snippet,
                 model_name_list, execution_time_list):
    """function to evalute the execution time of a code snippet.
    The results of the evalution will be saved in two lists.
    """

    start_time = time.perf_counter()
    eval(code_snippet)
    execution_time = time.perf_counter() - start_time
    model_name_list = model_name_list.append(name)
    execution_time_list = execution_time_list.append(execution_time)
    print(f"{name} Execution time: {execution_time}")



test_dict = {
    # Number of Stop words to remove
    "#Stopwords": [100, 2070],
    # Number of news articles to include
    "#news_articles": [200, 1000, 10000, 500000],
    #libraries and number of cores to test out
    "Libraries": {"Joblib": [2, 4, 8], "Ray":[2, 4],}
}

for number_of_texts in test_dict["#news_articles"]:
    # select the desidered number of articles for the experiment
    df = all_data.head(number_of_texts)
    for number_of_stopwords in test_dict["#Stopwords"]:

        stopwords_list = np.array(original_stopwords_list[:number_of_stopwords])
        # to store the results of the experiment
        model_name_list = []
        execution_time_list = []

        # sequential methods
        # from NLP_Preprocessing_Functions import remove_stop_words
        measure_time("Seq. List", "[remove_stop_words(x) for x in df['text'].to_list()]",
                     model_name_list, execution_time_list)
        measure_time("Seq. DF", "df['text'].apply(lambda x: remove_stop_words(x))",
                     model_name_list, execution_time_list)

        # multiprocessing methods
        for library in test_dict["Libraries"].keys():

            for n_core in test_dict["Libraries"][library]:
                if library == "Joblib":
                    statement = "Parallel(n_jobs=_CORES_)(delayed(remove_stop_words)(i) for i in df['text'].to_list())"
                else:
                    try:
                        pool = ray.util.multiprocessing.Pool(processes=nprocs)
                        statement = "pool.map(remove_stop_words, df['text'].to_list())"
                    except:
                        pass
                measure_time("Y XC".replace("X",
                            str(n_core)).replace("Y", library),
                            statement.replace("_CORES_", str(n_core)), model_name_list, execution_time_list)


        #plot the resutls
        plt.figure(figsize=(7,8), dpi=150)
        plt.style.use("seaborn")
        plt.bar(model_name_list, execution_time_list)
        plt.suptitle("Multiprocessing Strategies in Python", fontsize=22)
        plt.title("No. Texts: " + str(number_of_texts) +
                  ", No. Stopwords: " + str(number_of_stopwords), fontsize=18)
        plt.ylabel("Execution Time in Seconds", size=18)
        plt.xlabel("Multiprocessing Strategy", size=14)
        plt.axhline(min(execution_time_list), color="red")
        plt.xticks(rotation=20, size=18)
        plt.yticks(size=18)
        plt.savefig("./data/images/Multiprocessing_Experiment_Python_" +
                    str(number_of_texts) + "_articles_" +
                    str(number_of_stopwords) + "_stopwords.png")

        # store the results additionally as table

        # adding information on experiments
        execution_time_list = [number_of_texts, number_of_stopwords] + execution_time_list
        model_name_list = ["number_of_texts", "number_of_stopwords"] + model_name_list

        # check if dataframe exists
        if "results_df" in globals():
            new_data = pd.DataFrame(execution_time_list).transpose()
            new_data.columns = model_name_list
            results_df = results_df.append(new_data,
                              ignore_index=True)
        else:
            # create the dataframe
            results_df = pd.DataFrame(execution_time_list).transpose()
            results_df.columns = model_name_list

results_df.to_csv("./data/Preprocessing_Experiment_Results_Python.csv", sep=";", index=False)



